File: clojure/src/day2.clj
(ns day2
  (:require
   [clojure.string :refer [split-lines]]))

(defn is-valid-sequence [going-down]
  (fn [first, second]
    (let [distance (- first second)]
      (if going-down (< 0 distance 4) (> 0 distance -4)))))

(defn valid-plan [items]
  (every? true? (map (is-valid-sequence (> (first items) (second items))) items (rest items))))

(defn is-valid [line]
  (->> (re-seq #"\d+" line)
       (map parse-long)
       (valid-plan)))

(defn part1 [input]
  (->> (split-lines input)
       (map is-valid)
       (filter true?)
       (count)))

(defn find-combos [items]
  (as-> (range 0 (count items)) i
    (map #(concat (take % items) (drop (+ % 1) items)) i)))

(defn is-valid2 [line]
  (->> (re-seq #"\d+" line)
       (map parse-long)
       (find-combos)
       (map #(valid-plan %))
       (some true?)))

(defn part2 [input]
  (->> (split-lines input)
       (map is-valid2)
       (filter true?)
       (count)))

File: rust/src/day_2.rs
use itertools::Itertools;

pub fn check_line(lines: &[i32]) -> bool {
    let direction_up = lines[0] < lines[1];

    lines.windows(2).all(|window| {
        let [x, x2] = window.try_into().unwrap();
        let distance = (x - x2).abs();
        if distance > 3 {
            false
        } else if direction_up {
            x < x2
        } else {
            x > x2
        }
    })
}

pub fn a(input: &str) -> usize {
    input
        .lines()
        .filter(|line| {
            let items = line
                .split_whitespace()
                .map(|c| c.parse().unwrap())
                .collect_vec();

            check_line(&items)
        })
        .count()
}

pub fn b(input: &str) -> usize {
    input
        .lines()
        .filter(|line| {
            let items = line
                .split_whitespace()
                .map(|c| c.parse().unwrap())
                .collect_vec();

            (0..items.len()).any(|i| {
                let mut v = items.clone();
                v.remove(i);
                check_line(&v)
            })
        })
        .count()
}
